sctp/man/sctp.7
.\" t
.\" Don't change the line above, it tells man that tbl is needed.
.\" -*- nroff -*-
.\"
.\" @(#) $Id: sctp.7,v 0.9.2.19 2006/04/26 10:44:55 brian Exp $
.\"
.\" =========================================================================
.\"
.\" This manpage is Copyright (C) 2001-2002  OpenSS7 Corp. <www.openss7.com>
.\"
.\" All Rights Reserved.
.\"
.\" Permission is granted to make and distribute verbatim copies of this
.\" manual provided the copyright notice and this permission notice are
.\" preserved on all copies.
.\"
.\" Permission is granted to copy and distribute modified versions of this
.\" manual under the conditions for verbatim copying, provided that the
.\" entire resulting derived work is distributed under the terms of a
.\" permission notice identical to this one
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this
.\" manual page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from
.\" the use of the information contained herein.  The author(s) may not
.\" have taken the same level of care in the production of this manual,
.\" which is licensed free of charge, as they might when working
.\" professionally.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by
.\" the source, must acknowledge the copyright and authors of this work.
.\"
.\" =========================================================================
.\"
.\" Modified $Date: 2006/04/26 10:44:55 $ by $Author: brian $
.\"
.\" =========================================================================
.TH SCTP 7 "9 April 2003" "Linux Man Page" "Linux Programmer's Manual"
.SH "NAME"
sctp \- Stream Control Transmission Protocol (SCTP).
.SH "SYNOPSIS"
.nf
\fB#include <sys/socket.h>\fR

\fB#include <netinet/in.h>\fR

\fBsctp_socket = socket(PF_INET, \fISOCK_SEQPACKET\fB, 0);
\fBsctp_socket = socket(PF_INET, \fISOCK_SEQPACKET\fB, IPPROTO_SCTP);
\fBsctp_socket = socket(PF_INET, \fISOCK_STREAM\fB, IPPROTO_SCTP);
\fBsctp_socket = socket(PF_INET, \fISOCK_RDM\fB, IPPROTO_SCTP);
.fi

.SH "DESCRIPTION"
.PP
\fBSCTP\fR is an implementation of the SCTP (Stream Transmission Control Protocol) defined in
RFC 2960.  \fBSCTP\fR provides reliable delivery of ordered or unordered packets over a full
duplex connection between two \fI\fCSOCK_SEQPACKET\fR sockets on top of \fBip\fR(7).
\fBSCTP\fR can guarantee that the data arrives in order on a stream, if requested, and
retransmits lost packets.  It generates and checks a per packet checksum to catch transmission
errors.
.PP
\fBSCTP\fR provides the following socket types:
.TP
.B \f(CBSOCK_SEQPACKET\fR
A standard \fI\fCSOCK_SEQPACKET\fR socket that preserves message boundaries, Nagles at the
association level, and provides ordered and unordered reliable and partial reliable,
acknowledged and unacknowledged delivery on multiple streams.
.TP
.B \f(CBSOCK_STREAM\fR
A \fBtcp\fR(7) compatible \fI\fCSOCK_STREAM\fR socket that does not preserve message
boundaries, Nagles at the stream level, and provides ordered and out-of-band unacknowledged
delivery on a single stream.  This is an optional capability of \fBSCTP\fR which provides for
maximum compatibility with applications written for \fBtcp\fR(7).  This socket type is
available when \fBSCTP\fR is compiled with the kernel configuration parameter
\fI\fCCONFIG_SCTP_TCP_COMPATIBLE\fR set.
.TP
.B \f(CBSOCK_RDM\fR
A \fBudp\fR(7) compatible \fI\fCSOCK_RDM\fR socket that preserves message boundaries, does not
Nagle, and provides only unordered but reliable or partially reliable delivery on multiple
streams.  This is an optional capability of \fBSCTP\fR which provides for maximum compatibility
with applications written for \fBrudp\fR(7).  This socket type is available when \fBSCTP\fR is
compiled with the kernel configuration parameter \fI\fCCONFIG_SCTP_UDP_COMPATIBLE\fR set.
.SS "Connection Establishment"
.PP
A fresh \fBSCTP\fR socket has no remote or local address and is not fully specified.  To create
an outgoing \fBSCTP\fR association use \fBconnect\fR(2) to establish an association with
another \fBSCTP\fR endpoint.  To receive incoming associations \fBbind\fR(2) the socket first
to a number of local addresses and a port and then call \fBlisten\fR(2) to put the socket into
listening state.  Then a new socket for each incoming association can be accepted using
\fBaccept\fR(2).  A socket that has had \fBaccept\fR(2) or \fBconnect\fR(2) successfully called
on it is fully specified and may transmit data.  Data cannot be transmitted on listening or not
yet connected sockets.
.PP
\fBSCTP\fR supports large windows to support links with high latency or bandwidth.  Large
\fBSCTP\fR windows can be used by increasing the send and receive buffer sizes.  They can be
set globally with the \fBnet.core.wmem_default\fR and \fBnet.core.rmem_default\fR system
controls, or on individual sockets by using the \fI\fCSO_SNDBUF\fR and \fI\fCSO_RCVBUF\fR
socket options.  The maximum sizes for socket buffers are limited by the
\fBnet.core.wmem_max\fR and \fBnet.core.rmem_max\fR system controls.  See \fBsocket\fR(7) for
more information.
.SS "Sending Data"
.PP
When sending data, the specific destination address within the association to which to send the
data can be specified by providing a valid destination address as an argument to
\fBsendto\fR(2) or \fBsendmsg\fR(2).  Data sent with \fBsend\fR(2) or \fBwrite\fR(2) will use
\fBSCTP\fR's destination transport address selection policies for transmission and
retransmission to single- and multi-homed hosts.  The stream upon which to send the data can
also be set with the \fI\fCSCTP_SID\fR socket option on a socket basis with \fBsetsockopt\fR(2)
or on a per-message basis with \fI\fCSCTP_SID\fR control message to \fBsendmsg\fR(2).  In
addition, the \fBip\fR(7) \fI\fCIP_PKTINFO\fR option can be used with \fBsendmsg\fR(2) to
specify the interface, first-hop destination address and source address to be used in the
outbound packet on \fI\fCSOCK_SEQPACKET\fR and \fI\fCSOCK_RDM\fR sockets.  IP options can also
be specified for the outbound packet with the \fBip\fR(7) \fI\fCIP_OPTIONS\fR or
\fI\fCIP_RETOPTS\fR control messages to \fBsendmsg\fR(2).
.PP
For \fI\fCSOCK_SEQPACKET\fR socket, send operations may also use the \fI\fCMSG_MORE\fR flag
when sending with \fBsend\fR(2), \fBsendto\fR(2) or \fBsendmsg\fR(2) to indicate that the next
write operation will provide additional data belonging to the same packet.
\fI\fCSOCK_STREAM\fR sockets do not support send options and do not return the
\fI\fCMSG_TRUNC\fR flag on send operations.
.PP
If a \fI\fCSOCK_SEQPACKET\fR or \fI\fCSOCK_STREAM\fR socket does not have sufficient room to
buffer the sent data, it buffers what it can and returns the amount of user data buffered.
\fI\fCSOCK_RDM\fR sockets wait for sufficient room to buffer the entire sent packet before
returning.  \fI\fCSOCK_SEQPACKET\fR and \fI\fCSOCK_STREAM\fR sockets will segment large data
writes in to smaller segments for delivery to the peer.
.PP
\fBSCTP\fR provides a mechanism for receiving positive and negative acknowledgments of sent
data on \fI\fCSOCK_SEQPACKET\fR and \fI\fCSOCK_RDM\fR sockets.  When the socket option
\fI\fCSCTP_DISPOSITION\fR is set to \fI\fCSCTP_DISPOSITION_ACKED\fR, or when data is sent using
the \fI\fCMSG_CONFIRM\fR flag to \fBsend\fR(2), \fBsendto\fR(2) or \fBsendmsg\fR(2), message
confirmation or delivery failures can be received by passing the \fI\fCMSG_CONFIRM\fR flag to
\fBrecv\fR(2), \fBrecvfrom\fR(2) or \fBrecvmsg\fR(2).  This mechanism also provides for
retrieval and disposition of undelivered messages after the association has aborted or shut
down, but before a call to \fBclose\fR(2).
.SS "Receiving Data"
.PP
For \fI\fCSOCK_SEQPACKET\fR and \fI\fCSOCK_RDM\fR sockets, all receive operations return only
one packet, or a portion of one packet.  When the packet is smaller than the passed buffer only
that much data is returned, when it is larger the packet is truncated and the
\fI\fCMSG_TRUNC\fR flag is set for \fI\fCSOCK_RDM\fR sockets.  Subsequent \fBrecv\fR or
\fBread\fR(2), or a subsequent \fBrecvfrom\fR(2) or \fBrecvmsg\fR(2) for the same stream will
return the remainder of the packet for \fI\fCSOCK_SEQPACKET\fR sockets.  \fI\fCSOCK_SEQPACKET\fR
sockets also return the \fI\fCMSG_EOR\fR flag when the end of the packet has been received
successfully.  For \fI\fCSOCK_STREAM\fR sockets, all receive operations return only the amount
of data available on the current stream.  \fI\fCMSG_TRUNC\fR and \fI\fCMSG_EOR\fR are never set
for \fI\fCSOCK_STREAM\fR sockets.
.PP
IP options may be sent or received using the socket options described in \fBip\fR(7).  They are
only processed by the kernel when the appropriate system control is enabled (but still passed
to the user even when it is turned off).  See \fBip\fR(7).
.PP
When the \fI\fCMSG_DONTROUTE\fR flag is set on sending, the destination address must refer to a
local interface address and the packet is only sent to that interface.
.PP
\fBSCTP\fR fragments a packet when its total length exceeds the association MTU (Maximum
Transmission Unit).  A more network friendly alternative is to use path MTU discovery as
described in the \fI\fCIP_PMTU_DISCOVER\fR section of \fBip\fR(7).
.PP
\fBSCTP\fR supports urgent data similar to \fBtcp\fR(7).  Urgent data is used to signal the
receiver that some important message is part of the data stream and that it should be processed
as soon as possible.  Urgent data is always sent out of order.  To send urgent (or out of
order) data specify the \fI\fCMSG_OOB\fR option to \fBsend\fR(2), \fBsendto\fR(2), or
\fBsendmsg\fR(2).  When urgent data is received, the kernel sends a \fI\fCSIGURG\fR signal to
the reading process or the process or process group that has been set for the socket using the
\fI\fCFIOCSPGRP\fR or \fI\fCFIOCSETOWN\fR I/O controls.  When the \fI\fCSO_OOBINLINE\fR socket
option is enabled, urgent data is put into the normal data stream (and can be tested for by the
\fI\fCSIOCATMARK\fR I/O control), otherwise it can only be received when the \fI\fCMSG_OOB\fR
flag is set for \fBrecvmsg\fR(2).
.SS "ADDRESS FORMATS"
\fBSCTP\fR is built on top of \fBIP\fR (see \fBip\fR(7)).  The address formats defined by
\fBip\fR(7) apply to \fBSCTP\fR.  \fBSCTP\fR uses the IPv4 \fBsockaddr_in\fR address format
described in \fBip\fR(7).
.PP
\fBSCTP\fR supports multiple-stream point-to-point communication within multi-homed
associations; broadcasting and multi-casting are not supported.
.PP
In addition to normal IP addressing, \fBSCTP\fR provides extensions on the \fBbind\fR(2),
\fBaccept\fR(2), \fBconnect\fR(2), \fBgetsockname\fR(2), and \fBgetpeername\fR(2) system calls.
These functions normally take or return a single \fBsockaddr_in\fR as an address argument.
When used in conjunction with an \fBSCTP\fR socket on or to a multi-homed host, these calls
will accept or return an array of \fBsockaddr_in\fR structures.  The number of addresses in the
structure is indicated by the length of the structure and the protocol family of the socket.
\fBSCTP\fR will only support one port number being specified in the address list, and each
\fBsockaddr_in\fR structure in the list must contain the same \fIsin_port\fR.  (Note: Linux
uses \fI\fCMAX_SOCK_ADDR\fR as a maximum size of the socket address length.  This is currently
only 128 bytes or eight (8) \fBsockaddr_in\fR structures. Eight (8) addresses should be
sufficient for most applications.)
.PP
Multiple addresses provided to the \fBbind\fR(2) system call will be interpreted as multiple
local addresses to provide to the peer at connection time as well as the local address/port
combinations upon which a listening socket will accept incoming associations.  A call to
\fBaccept\fR(2) requesting the source address of the peer will return multiple transport
addresses if the connecting peer is multi-homed.  Multiple addresses provided to the
\fBconnect\fR(2) system call will be interpreted as multiple destination address/port
combinations to which to attempt to form an association. The \fBgetpeername\fR(2) system call
returns the list of destination address/port combinations to which the socket is connected.
The \fBgetsockname\fR(2) system call returns the list of local address/port combinations to
which the socket is bound.
.TP
.BR getsockname (2)
A call to \fBgetsockname\fR(2) can provide multiple bound transport addresses for a multi-homed
host when the socket is bound to multiple addresses or \fI\fCINADDR_ANY\fR using \fBbind\fR(2).
The first address in the list is the primary address to which the socket is bound.  This is the
first address that will be attempted to be used as a source address when sending an INIT chunk
to the \fBSCTP\fR peer as a result of a call to \fBconnect\fR(2).  Otherwise, the order of the
addresses has no significance.  When the socket is not bound to any address, the call to
\fBgetsockname\fR(2) returns a socket address with address family \fI\fCAF_UNSPEC\fR.
.IP
As a result of binding to \fI\fCINADDR_ANY\fR or the \fBADD-IP\fR extension, subsequent calls
to \fBgetsockname\fR(2) may return different addresses depending on the state of the connection
or resulting from the exchange of ASCONF control chunks.  This list of transport addresses
returned by \fBgetsockname\fR(2) represents the current list of local transport addresses bound
to the endpoint.  When the connection is multi-homed or supports \fBADD-IP\fR, this behavior is
different from \fBtcp\fR(7) and other connection-oriented protocols, whose bound address does
not change during the lifespan of a connection.
.TP
.BR getpeername (2)
A call to \fBgetpeername\fR(2) can provide multiple peer transport addresses for a multi-homed
peer when the socket is connecting or connected to multiple addresses using \fBconnect\fR(2).
The first address is always the current primary destination transport address.  The current
primary destination transport address is the address to which all messages will be sent in the
absence of congestion or failure.
.IP
As a result of connecting to multiple transport addresses using \fBconnect\fR(2) or the
\fBADD-IP\fR extension, the membership of the list of transport addresses returned by
\fBgetpeername\fR(2) may change due to the state of the socket or resulting from the exchange
of ASCONF control chunks.  When the \fBSCTP\fR connection is multi-homed or supports
\fBADD-IP\fR, this behavior is different from \fBtcp\fR(7) and other connection-oriented
protocol, whose peer address does not change during the lifespan of a connection.
.SS "SOCKET CALLS"
.PP
Socket calls for \fI\fCSOCK_STREAM\fR, \fI\fCSOCK_SEQPACKET\fR and \fI\fCSOCK_RDM\fR sockets
are different in some respects (addressing and options) than their \fBtcp\fR(7) or \fBudp\fR(7)
counterparts.  These differences are described here.  For common behavior, please see the
indicated manual page in Section 2.
.TP
.BR bind (2)
A call to \fBbind\fR(2) can provide multiple address if the host is multi-homed.  Binds to
\fI\fCINADDR_ANY\fR will result in a binding to all of the local transport addresses belonging
to the host that do not cause a conflict at the time that a connection is formed.  As with
other sockets, the local port number is assigned at the time that \fBbind\fR(2) is called.
.IP
As with other sockets, binding to a socket address with address family \fI\fCAF_UNSPEC\fR will
cause the socket to be unbound from all transport addresses.
.IP
Consistent with \fBtcp\fR(7), \fBSCTP\fR supports binding of multiple sockets to the same
transport address providing that the socket option \fI\fCSO_REUSEADDR\fR (see \fBsocket\fR(7))
is set on the sockets and that no sockets have executed \fBlisten\fR(2) bound to the same
transport address.
.TP
.BR accept (2)
A call to \fBaccept\fR(2) requesting the peer address of the connection will return multiple
transport addresses if the connecting peer is multi-homed.  The first transport address in the
list is the primary transport address of the connecting peer.  The primary transport address is
the address upon which the INIT message was received.  Otherwise, the \fBaccept\fR(2) operation
is unchanged.
.TP
.BR connect (2)
A call to \fBconnect\fR(2) can provide multiple destination transport addresses if the peer
host is multi-homed.  If an INIT chunk sent to the first destination transport address is not
acknowledged, the next INIT chunk will be sent to the next address in the list.  This will
occur until an acknowledgment is received from one of the addresses or the connection attempt
times out.  Each address will be retried \fBsctp_max_init_retries\fR times before the
connection process is considered to have timed out.  Broadcast and multicast addresses are
permitted for the call to \fBconnect\fR(2) but will not form an endpoint address in the
association.  This feature permits fast and reliable initialization of associations.
.IP
As with other connection-oriented sockets, connecting to a socket address with address family
\fI\fCAF_UNSPEC\fR will cause a socket engaged in active connections to be disconnected.
.TP
.BR read "(2), " recv "(2), " recvfrom "(2) and " recvmsg "(2)"
\fBSCTP\fR preserves message boundaries on read for \fI\fCSOCK_SEQPACKET\fR and
\fI\fCSOCK_RDM\fR sockets.  Message boundaries are never preserved for \fI\fCSOCK_STREAM\fR
sockets.
.IP
The following message flags to, or returned from, \fBrecv\fR(2), \fBrecvfrom\fR(2) and
\fBrecvmsg\fR(2) have special interpretations for \fBSCTP\fR:
.RS
.TP
.B \f(CBMSG_OOB
\fBSCTP\fR supports the return of the \fI\fCMSG_OOB\fR flag from \fBrecv\fR(2),
\fBrecvfrom\fR(2) and \fBrecvmsg\fR(2).  Data read with the \fI\fCMSG_OOB\fR flag set indicates
that the data was received on the specified \fBSCTP\fR stream with the Unordered bit set.
Operation is similar to \fBtcp\fR(7).
.TP
.B \f(CBMSG_EOR
For \fI\fCSOCK_SEQPACKET\fR sockets, the \fI\fCMSG_EOR\fR flag returned from \fBrecv\fR(2),
\fBrecvfrom\fR(2) or \fBrecvmsg\fR(2) indicates that the data read has been read to the end of
a record.  The \fI\fCMSG_EOR\fR flag is not used for \fI\fCSOCK_STREAM\fR or \fI\fCSOCK_RDM\fR
sockets.
.TP
.B \f(CBMSG_TRUNC
\fBSCTP\fR supports the use of the \fI\fCMSG_TRUNC\fR flag to \fBrecv\fR(2), \fBrecvfrom\fR(2)
and \fBrecvmsg\fR(2).  Data read with the \fI\fCMSG_TRUNC\fR flag set will return the number of
bytes available in the packet rather than the number of bytes read.
.IP
When the \fI\fCMSG_TRUNC\fR flag is returned from \fBrecv\fR(2), \fBrecvfrom\fR(2) and
\fBrecvmsg\fR(2), it indicates that the record was truncated.  \fI\fCMSG_TRUNC\fR will only be
returned on \fI\fCSOCK_RDM\fR sockets.  \fI\fCMSG_TRUNC\fR is never set on return for
\fI\fCSOCK_STREAM\fR and \fI\fCSOCK_SEQPACKET\fR sockets.
.TP
.B \f(CBMSG_CONFIRM
\fBSCTP\fR supports the use of the \fI\fCMSG_CONFIRM\fR flag to \fBrecv\fR(2),
\fBrecvfrom\fR(2) and \fBrecvmsg\fR(2).  Data read with the \fI\fCMSG_CONFIRM\fR flag set will
return messages receipt acknowledgments as well as messages which have exceeded their
life-times, messages which have been dropped by \fBPR-SCTP\fR, and messages held for retrieval
after abort or shutdown.
.IP
When the \fI\fCMSG_CONFIRM\fR flag is returned from \fBrecv\fR(2), \fBrecvfrom\fR(2) and
\fBrecvmsg\fR(2), it indicates that the read data represents data that was confirmed as
acknowledged by the peer.  See also \fI\fCSCTP_DISPOSITION\fR under \fBSOCKET OPTIONS\fR.
.RE
.IP
Other message flags have the same interpretation as described in \fBrecv\fR(2),
\fBrecvfrom\fR(2) and \fBrecvmsg\fR(2).
.TP
.BR write "(2), " send "(2), " sendto "(2) and " sendmsg "(2)"
\fBSCTP\fR preserves message boundaries on write for \fI\fCSOCK_SEQPACKET\fR and
\fI\fCSOCK_RDM\fR sockets.  Message boundaries are never preserved for \fI\fCSOCK_STREAM\fR
sockets.
.IP
The following message flags to \fBsend\fR(2), \fBsendto\fR(2) and \fBsendmsg\fR(2) have special
interpretations for \fBSCTP\fR:
.RS
.TP
.B \f(CBMSG_OOB
\fBSCTP\fR supports the use of the \fI\fCMSG_OOB\fR flag to \fBsend\fR(2), \fBsendto\fR(2) and
\fBsendmsg\fR(2).  Data written with the \fI\fCMSG_OOB\fR flag set indicates that the data is
to be sent on the specified \fBSCTP\fR stream with the Unordered bit set.
.TP
.B \f(CBMSG_MORE
\fBSCTP\fR supports the use of the \fI\fCMSG_MORE\fR flag to \fBsend\fR(2), \fBsendto\fR(2) and
\fBsendmsg\fR(2).  Data written with the \fI\fCMSG_MORE\fR flag set indicates that the data of
a subsequent send operation on the same stream contains additional data belonging to the same
record.  Use of the \fI\fCMSG_MORE\fR flag is only supported for \fI\fCSOCK_SEQPACKET\fR
sockets.  Data written with \fBwrite\fR(2) is assumed to contain an entire record.
.TP
.B \f(CBMSG_PROBE
\fBSCTP\fR supports the use of the \fI\fCMSG_PROBE\fR flag to \fBsend\fR(2), \fBsendto\fR(2)
and \fBsendmsg\fR(2).  Data written with the \fI\fCMSG_PROBE\fR flag set indicates that the
data is to be used to fill out the heartbeat data in a HEARTBEAT chunk and request that
\fBSCTP\fR send a HEARTBEAT to the peer.  Use of the \fI\fCMSG_PROBE\fR flag is supported on
all socket types.
.TP
.B \f(CBMSG_EOF
\fBSCTP\fR supports the use of the \fI\fCMSG_EOF\fR flag to \fBsend\fR(2), \fBsendto\fR(2) and
\fBsendmsg\fR(2).  Data written with the \fI\fCMSG_EOF\fR flag will be the last data sent on
the association and then a shutdown initiated.
.TP
.B \f(CBMSG_CONFIRM
\fBSCTP\fR supports the use of the \fI\fCMSG_CONFIRM\fR flag to \fBsend\fR(2) \fBsendto\fR(2)
and \fBsendmsg\fR(2) for \fI\fCSOCK_SEQPACKET\fR and \fI\fCSOCK_RDM\fR sockets.  Records or
packets written with the \fI\fCMSG_CONFIRM\fR flag set are marked for acknowledgment.
Acknowledgments can be received with the \fI\fCMSG_CONFIRM\fR flag to \fBrecv\fR(2),
\fBrecvfrom\fR(2) and \fBrecvmsg\fR(2) as described above.  See also \fI\fCSCTP_DISPOSITION\fR
under \fBSOCKET OPTIONS\fR.
.RE
.IP
Other message flags have the same interpretation as described in \fBsend\fR(2), \fBsendto\fR(2)
and \fBsendmsg\fR(2).
.TP
.BR shutdown "(2) and " close (2)

NEWS
OpenSS7 Master Package -- history of user-visible changes.  2006-03-23
$Id: NEWS,v 0.9.2.12 2006/03/24 00:04:20 brian Exp $
Copyright (c) 2001-2006  OpenSS7 Corporation. <http://www.openss7.com/>
Copyright (c) 1997-2000  Brian Bidulock <bidulock@openss7.org>
See the end for copying conditions.

* Major changes in openss7-0.9.2.D

  Added --enable-devel configure option for embedded targets.
  Added send-pr script for automatic problem report generation.

* Major changes in openss7-0.9.2.C

  Distribution check for entire master package.  Trying to get master
  package into form where it can be released as a complete package.

* Major changes in openss7-0.9.2.B

  Minor changes for wider release, better master packaging and bug
  fixes.

* Initial release openss7-0.9.2.A

  With this release version numbers were changed to reflect an upstream
  version only to be consistent with other OpenSS7 package releases.
  All RPM release numbers will be -1$(PACKAGE_RPMEXTRA) and all Debian
  release numbers will be _0.  If you wish to apply patches and
  relelease the package, please bump up the release number and apply a
  suitable release suffix for your organization.  We leave Debian
  release number _1 reserved for your use, so you can still bundle the
  source in the .dsc file.

  Major changes for this release include build against Linux 2.6 kernels
  and popular distributions based on the 2.6 kernel as well as wider
  distribution support.

* Initial release openss7-0.9.2-1

  Initial autoconf/RPM release of the OpenSS7 master package.

  This master package contains all other OpenSS7 releases.


=========================================================================

Copyright (c) 2001-2006  OpenSS7 Corporation. <http://www.openss7.com/>
Copyright (c) 1997-2000  Brian Bidulock <bidulock@openss7.org>

All Rights Reserved.

 Permission is granted to make and distribute verbatim copies of this
 manual provided the copyright notice and this permission notice are
 preserved on all copies.

 Permission is granted to copy and distribute modified versions of this
 manual under the conditions for verbatim copying, provided that the
 entire resulting derived work is distributed under the terms of a
 permission notice identical to this one

 Since the Linux kernel and libraries are constantly changing, this
 manual page may be incorrect or out-of-date.  The author(s) assume no
 responsibility for errors or omissions, or for damages resulting from
 the use of the information contained herein.  The author(s) may not
 have taken the same level of care in the production of this manual,
 which is licensed free of charge, as they might when working
 professionally.

 Formatted or processed versions of this manual, if unaccompanied by
 the source, must acknowledge the copyright and authors of this work.

-------------------------------------------------------------------------

 U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software
 on behalf of the U.S. Government ("Government"), the following
 provisions apply to you.  If the Software is supplied by the Department
 of Defense ("DoD"), it is classified as "Commercial Computer Software"
 under paragraph 252.227-7014 of the DoD Supplement to the Federal
 Acquisition Regulations ("DFARS") (or any successor regulations) and the
 Government is acquiring only the license rights granted herein (the
 license rights customarily provided to non-Government users).  If the
 Software is supplied to any unit or agency of the Government other than
 DoD, it is classified as "Restricted Computer Software" and the
 Government's rights in the Software are defined in paragraph 52.227-19
 of the Federal Acquisition Regulations ("FAR") (or any successor
 regulations) or, in the cases of NASA, in paragraph 18.52.227-86 of the
 NASA Supplement to the FAR (or any successor regulations).

=========================================================================

 Commercial licensing and support of this software is available from
 OpenSS7 Corporation at a fee.  See http://www.openss7.com/

=========================================================================
vim: ft=README tw=72 nocindent nosmartindent formatoptions+=tcqlorn


sctp/util/init/sctp.sh
#!/bin/sh
#
# @(#) $RCSfile: sctp.sh,v $ $Name:  $($Revision: 0.9.2.1 $) $Date: 2006/03/29 11:27:59 $
# Copyright (c) 2001-2006  OpenSS7 Corporation <http://www.openss7.com/>
# Copyright (c) 1997-2000  Brian F. G. Bidulock <bidulock@openss7.org>
# All Rights Reserved.
#
# Distributed by OpenSS7 Corporation.  See the bottom of this script for copying
# permissions.
#
# These are arguments to update-rc.d ala chkconfig and lsb.  They are recognized
# by openss7 install_initd and remove_initd scripts.  Each line specifies
# arguments to add and remove links after the the name argument:
#
# sctp:		start and stop SCTP subsystem
# update-rc.d:	start 33 S . stop 33 0 6 .
# config:	/etc/default/sctp
# probe:	false
# hide:		false
# license:	GPL
# description:	This init script is part of OpenSS7 Linux SCTP.  \
#		It is responsible for ensuring that the SCTP subsystem is \
#		configured and loaded.
#

PATH=/usr/local/sbin:/usr/local/bin:/sbin:/bin:/usr/sbin:/usr/bin
name='sctp'
config="/etc/default/$name"
desc="the SCTP subsystem"

[ -e /proc/modules ] || exit 0

for SCTP_MKNOD in /sbin/${name}_mknod /usr/sbin/${name}_mknod /bin/${name}_mknod /usr/bin/${name}_mknod ; do
    if [ -x $SCTP_MKNOD ] ; then
	break
    else
	SCTP_MKNOD=
    fi
done

# Specify defaults

SCTP_MODULES="sctp"
SCTP_MAKEDEVICES="no"
SCTP_REMOVEDEVICES="no"

# Source config file
for file in $config ; do
    [ -f $file ] && . $file
done

[ -z "$SCTP_MKNOD" ] && SCTP_MAKEDEVICES='no'
[ -z "$SCTP_MKNOD" ] && SCTP_REMOVEDEVICES='no'

RETVAL=0

umask 077

if [ "${VERBOSE:-0}" -ne 0 ] ; then
    redir='>/dev/null 2>&1'
else
    redir=
fi

build_options() {
    # Build up the options string
    :
}

start() {
    echo -n "Loading SCTP kernel modules: "
    for module in $SCTP_MODULES ; do
	if ! grep "^$module"'[[:space:]]' /proc/modules $redir ; then
	    echo -n "$module "
	    modprobe -k -q -- $module $redir
	    [ $? -eq 0 ] || echo -n "(failed)"
	fi
    done
    echo "."

    echo -n "Starting $desc: $name "
    build_options
    RETVAL=$?
    if [ $RETVAL -eq 0 ] ; then
	echo "."
    else
	echo "(failed.)"
    fi
    if grep '^[[:space:]]*'${name}'[/.]' /etc/sysctl.conf $redir ; then
	echo -n "Reconfiguring kernel parameters: "
	sysctl -p /etc/sysctl.conf $redir
	RETVAL=$?
	if [ $RETVAL -eq 0 ] ; then
	    echo "."
	else
	    echo "(failed.)"
	fi
    fi
    if [ -f /etc/${name}.conf ] ; then
	echo -n "Configuring SCTP parameters: "
	sysctl -p /etc/${name}.conf $redir
	RETVAL=$?
	if [ $RETVAL -eq 0 ] ; then
	    echo "."
	else
	    echo "(failed.)"
	fi
    fi
    if [ -n "$STRINET_MKNOD" -a ":$STRINET_MAKEDEVICES" = ":yes" ] ; then
	echo -n "Making SCTP devices: "
	$SCTP_MKNOD
	RETVAL=$?
	if [ $RETVAL -eq 0 ] ; then
	    echo "."
	else
	    echo "(failed.)"
	fi
    fi
    return $RETVAL
}

remove_modules() {
    modules=
    while read -a module ; do
	modules="${modules}${modules:+ }${module[0]}"
    done
    if [ -n "$modules" ] ; then
	echo -n "Removing SCTP modules: "
	rmmod $modules
	RETVAL=$?
    fi
    return $RETVAL
}

stop() {
    echo -n "Stopping $desc: $name "
    RETVAL=$?
    if [ -n "$SCTP_MKNOD" -a ":$SCTP_REMOVEDEVICES" = ":yes" ] ; then
	echo -n "Removing SCTP devices: "
	$SCTP_MKNOD --remove
	RETVAL=$?
    fi
    [ $RETVAL -eq 0 ] && egrep '^sctp' /proc/modules 2>/dev/null | remove_modules
    RETVAL=$?
    if [ $RETVAL -eq 0 ] ; then
	echo "."
    else
	echo "(failed.)"
    fi
    return $RETVAL
}

restart() {
    stop
    start
    return $?
}

show() {
    echo "$name.sh: show: not yet implemented." >&2
    return 1
}

usage() {
    echo "Usage: /etc/init.d/$name.sh (start|stop|restart|force-reload|show)" >&2
    return 1
}

case "$1" in
    (start|stop|restart|show)
	$1 || RETVAL=$?
	;;
    (force-reload)
	restart || RETVAL=$?
	;;
    (*)
	usage || RETVAL=$?
	;;
esac

[ "${0##*/}" = "$name.sh" ] && exit $RETVAL

# =============================================================================
# 
# @(#) $RCSfile: sctp.sh,v $ $Name:  $($Revision: 0.9.2.1 $) $Date: 2006/03/29 11:27:59 $
#
# -----------------------------------------------------------------------------
#
# Copyright (c) 2001-2006  OpenSS7 Corporation <http://www.openss7.com/>
# Copyright (c) 1997-2000  Brian F. G. Bidulock <bidulock@openss7.org>
#
# All Rights Reserved.
#
# This program is free software; you can redistribute it and/or modify it under
# the terms of the GNU General Public License as published by the Free Software
# Foundation; version 2 of the License.
#
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
# details.
#
# You should have received a copy of the GNU General Public License along with
# this program; if not, write to the Free Software Foundation, Inc., 675 Mass
# Ave, Cambridge, MA 02139, USA.
#
# -----------------------------------------------------------------------------
#
# U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
# behalf of the U.S. Government ("Government"), the following provisions apply
# to you.  If the Software is supplied by the Department of Defense ("DoD"), it
# is classified as "Commercial Computer Software" under paragraph 252.227-7014
# of the DoD Supplement to the Federal Acquisition Regulations ("DFARS") (or any
# successor regulations) and the Government is acquiring only the license rights
# granted herein (the license rights customarily provided to non-Government
# users).  If the Software is supplied to any unit or agency of the Government
# other than DoD, it is classified as "Restricted Computer Software" and the
# Government's rights in the Software are defined in paragraph 52.227-19 of the
# Federal Acquisition Regulations ("FAR") (or any successor regulations) or, in
# the cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the FAR
# (or any successor regulations).
#
# -----------------------------------------------------------------------------
#
# Commercial licensing and support of this software is available from OpenSS7
# Corporation at a fee.  See http://www.openss7.com/
#
# -----------------------------------------------------------------------------
#
# Last Modified $Date: 2006/03/29 11:27:59 $ by $Author: brian $
#
# =============================================================================

# vim: ft=sh sw=4 tw=80
